core/filtermaps/map_renderer.go

```diff
@@ -257,10 +256,13 @@ func (f *FilterMaps) loadHeadSnapshot() error {
 
 // makeSnapshot creates a snapshot of the current state of the rendered map.
 func (r *mapRenderer) makeSnapshot() {
-	r.f.renderSnapshots.Add(r.iterator.blockNumber, &renderedMap{
-		filterMap:     r.currentMap.filterMap.copy(),
+	if r.iterator.blockNumber != r.currentMap.lastBlock {
+		panic("iterator state inconsistent with last block")
+	}
+	r.f.renderSnapshots.Add(r.currentMap.lastBlock, &renderedMap{
+		filterMap:     r.currentMap.filterMap.fastCopy(),
 		mapIndex:      r.currentMap.mapIndex,
-		lastBlock:     r.iterator.blockNumber,
+		lastBlock:     r.currentMap.lastBlock,
 		lastBlockId:   r.f.targetView.getBlockId(r.currentMap.lastBlock),
```

Comment:
Honestly it's a bit weird to get the block id with `r.f.targetView`, `r.iterator.chainView` might be a better candidate.

Reply:
You are right, it does look better. It is an assumption that the two should always be the same, guaranteed by `iterator.updateChainView` that is always called after `r.f.targetView` could have been changed. But it is better to explicitly check rather that quietly assume so I also added a check.